File: utbot-framework/src/main/java/org/utbot/quickcheck/generator/Generator.java


package org.utbot.quickcheck.generator;

import org.javaruntype.type.TypeParameter;
import org.javaruntype.type.Types;
import org.javaruntype.type.WildcardTypeParameter;
import org.utbot.quickcheck.internal.ReflectionException;
import org.utbot.quickcheck.random.SourceOfRandomness;

import java.lang.annotation.Annotation;
import java.lang.reflect.AnnotatedElement;
import java.lang.reflect.AnnotatedType;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

import static java.util.Collections.*;
import static java.util.stream.Collectors.toList;
import static org.utbot.quickcheck.internal.Reflection.*;

/**
 * Produces values for property parameters.
 *
 * @param <T> type of property parameter to apply this generator's values to
 */
public abstract class Generator<T> implements Gen<T> {
    private final List<Class<T>> types = new ArrayList<>();

    private Generators repo;

    /**
     * @param type class token for type of property parameter this generator is
     * applicable to
     */
    protected Generator(Class<T> type) {
        this(singletonList(type));
    }

    /**
     * Used for generators of primitives and their wrappers. For example, a
     * {@code Generator<Integer>} can be used for property parameters of type
     * {@code Integer} or {@code int}.
     *
     * @param types class tokens for type of property parameter this generator
     * is applicable to
     */
    protected Generator(List<Class<T>> types) {
        this.types.addAll(types);
    }

    /**
     * @return class tokens for the types of property parameters this generator
     * is applicable to
     */
    public List<Class<T>> types() {
        return unmodifiableList(types);
    }

    /**
     * Tells whether this generator is allowed to be used for property
     * parameters of the given type.
     *
     * @param type type against which to test this generator
     * @return {@code true} if the generator is allowed to participate in
     * generating values for property parameters of {@code type}
     */
    public boolean canRegisterAsType(Class<?> type) {
        return true;
    }

    /**
     * <p>This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.</p>
     *
     * @return whether this generator has component generators, such as for
     * those generators that produce lists or maps
     * @see #addComponentGenerators(List)
     */
    public boolean hasComponents() {
        return false;
    }

    /**
     * <p>This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.</p>
     *
     * @return how many component generators this generator needs
     * @see #addComponentGenerators(List)
     */
    public int numberOfNeededComponents() {
        return 0;
    }

    /**
     * <p>Adds component generators to this generator.</p>
     *
     * <p>Some generators need component generators to create proper values.
     * For example, list generators require a single component generator in
     * order to generate elements that have the type of the list parameter's
     * type argument.</p>
     *
     * <p>This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.</p>
     *
     * @param newComponents component generators to add
     */
    public void addComponentGenerators(List<Generator<?>> newComponents) {
        // do nothing by default
    }

    /**
     * @param typeParameters a list of generic type parameters
     * @return whether this generator can be considered for generating values
     * for property parameters that have the given type parameters in their
     * signatures
     */
    public boolean canGenerateForParametersOfTypes(List<TypeParameter<?>> typeParameters) {
        return true;
    }

    /**
     * <p>Configures this generator using annotations from a given annotated
     * type.</p>
     *
     * <p>This method considers only annotations that are themselves marked
     * with {@link GeneratorConfiguration}.</p>
     *
     * <p>By default, the generator will configure itself using this
     * procedure:</p>
     * <ul>
     *   <li>For each of the given annotations:
     *     <ul>
     *       <li>Find a {@code public} method on the generator named
     *       {@code configure}, that accepts a single parameter of the
     *       annotation type</li>
     *       <li>Invoke the {@code configure} method reflectively, passing the
     *       annotation as the argument</li>
     *     </ul>
     *   </li>
     * </ul>
     *
     * @param annotatedType a type usage
     * @throws GeneratorConfigurationException if the generator does not
     * "understand" one of the generation configuration annotations on
     * the annotated type
     */
    public void configure(AnnotatedType annotatedType) {
        configureStrict(collectConfigurationAnnotations(annotatedType));
    }

    /**
     * @param element an annotated program element
     */
    public void configure(AnnotatedElement element) {
        configureLenient(collectConfigurationAnnotations(element));
    }

    /**
     * <p>Supplies the available generators to this one.</p>
     *
     * <p>This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.</p>

     * @param provided repository of available generators
     */
    public void provide(Generators provided) {
        repo = provided;
    }

    /**
     * Used by the framework to make a copy of the receiver.
     *
     * @return a copy of the receiver
     */
    @SuppressWarnings("unchecked")
    public Generator<T> copy() {
        return (Generator<T>) instantiate(getClass());
    }

    /**
     * @param parameter a generic type parameter
     * @param clazz a type
     * @return whether the type is compatible with the generic type parameter
     * @see #canGenerateForParametersOfTypes(List)
     */
    protected static boolean compatibleWithTypeParameter(
        TypeParameter<?> parameter,
        Class<?> clazz) {

        return parameter instanceof WildcardTypeParameter
            || parameter.getType().isAssignableFrom(Types.forJavaLangReflectType(clazz));
    }

    /**
     * @return an access point for the available generators
     */
    protected Generators gen() {
        return repo;
    }

    /**
     * @param random a source of randomness used when locating other
     * generators
     * @return an access point for the available generators
     */
    protected Generators gen(SourceOfRandomness random) {
        return repo.withRandom(random);
    }

    /**
     * Gives a list of the {@link GeneratorConfiguration} annotations present
     * on the given program element.
     *
     * @param element an annotated program element
     * @return what configuration annotations are present on that element
     */
    protected static List<Annotation> configurationAnnotationsOn(AnnotatedElement element) {
        return allAnnotations(element).stream()
            .filter(a -> a.annotationType().isAnnotationPresent(GeneratorConfiguration.class))
            .collect(toList());
    }

    private Map<Class<? extends Annotation>, Annotation> collectConfigurationAnnotations(
        AnnotatedElement element) {

        if (element == null)
            return emptyMap();

        List<Annotation> configs = configurationAnnotationsOn(element);

        Map<Class<? extends Annotation>, Annotation> byType = new HashMap<>();
        for (Annotation each : configs)
            byType.put(each.annotationType(), each);
        return byType;
    }

    private void configureStrict(Map<Class<? extends Annotation>, Annotation> byType) {
        for (Map.Entry<Class<? extends Annotation>, Annotation> each : byType.entrySet())
            configureStrict(each.getKey(), each.getValue());
    }

    private void configureStrict(
        Class<? extends Annotation> annotationType,
        Annotation configuration) {

        configure(
            annotationType,
            configuration,
            ex -> {
                throw new GeneratorConfigurationException(
                    String.format(
                        "Generator %s does not understand configuration annotation %s",
                        getClass().getName(),
                        annotationType.getName()),
                    ex);
            });
    }

    private void configureLenient(Map<Class<? extends Annotation>, Annotation> byType) {
        for (Map.Entry<Class<? extends Annotation>, Annotation> each : byType.entrySet())
            configureLenient(each.getKey(), each.getValue());
    }

    private void configureLenient(
        Class<? extends Annotation> annotationType,
        Annotation configuration) {

        configure(annotationType, configuration, ex -> {});
    }

    private void configure(
        Class<? extends Annotation> annotationType,
        Annotation configuration,
        Consumer<ReflectionException> exceptionHandler) {

        Method configurer = null;

        try {
            configurer = findMethod(getClass(), "configure", annotationType);
        } catch (ReflectionException ex) {
            exceptionHandler.accept(ex);
        }

        if (configurer != null)
            invoke(configurer, this, configuration);
    }
}

File: utbot-framework/src/main/kotlin/org/utbot/quickcheck/generator/Generator.kt
package org.utbot.quickcheck.generator

import org.javaruntype.type.TypeParameter
import org.javaruntype.type.Types
import org.javaruntype.type.WildcardTypeParameter
import org.utbot.engine.greyboxfuzzer.generator.GeneratorConfigurator
import org.utbot.engine.greyboxfuzzer.util.FuzzerIllegalStateException
import org.utbot.engine.greyboxfuzzer.util.getImplementersOfWithChain
import org.utbot.engine.greyboxfuzzer.util.removeIfAndReturnRemovedElements
import org.utbot.framework.plugin.api.UtModel
import org.utbot.framework.plugin.api.UtNullModel
import org.utbot.framework.plugin.api.util.objectClassId
import org.utbot.quickcheck.internal.Reflection
import org.utbot.quickcheck.internal.ReflectionException
import org.utbot.quickcheck.random.SourceOfRandomness
import soot.SootClass
import java.lang.reflect.AnnotatedElement
import java.lang.reflect.AnnotatedType
import java.lang.reflect.Method
import java.util.Collections
import java.lang.annotation.Annotation as JavaAnnotation

/**
 * Produces values for property parameters.
 *
 */
abstract class Generator protected constructor(types: List<Class<*>>) : Gen {
    private val types: MutableList<Class<*>> = ArrayList()
    private var repo: Generators? = null
    var generatedUtModel: UtModel? = null
    var generationState = GenerationState.REGENERATE
    var nestedGenerators = mutableListOf<Generator>()
    lateinit var generatorContext: GeneratorContext

    open fun generateImpl(random: SourceOfRandomness, status: GenerationStatus): UtModel {
        if (generatorContext.checkPoint()) {
            return UtNullModel(objectClassId)
        }
        return when (generationState) {
            GenerationState.REGENERATE -> {
                generate(random, status).also {
                    generatedUtModel = it
                    nestedGeneratorsRecursive().forEach {
                        it.generationState = GenerationState.CACHE
                    }
                }
            }
            GenerationState.CACHE -> {
                generatedUtModel ?: throw FuzzerIllegalStateException("No cached model")
            }
            GenerationState.MODIFY -> {
                withModification {
                    generate(random, status).also { generatedUtModel = it }
                }
            }
            GenerationState.MODIFYING_CHAIN -> {
                generate(random, status).also {
                    generatedUtModel = it
                    nestedGeneratorsRecursive().forEach {
                        it.generationState = GenerationState.CACHE
                    }
                }
            }
        }
    }

    private fun flattenedTo(destination: MutableList<Generator>) {
        destination.add(this)
        nestedGenerators.forEach { it.flattenedTo(destination) }
    }

    fun nestedGeneratorsRecursive(): List<Generator> {
        val allGenerators = mutableListOf<Generator>()
        this.flattenedTo(allGenerators)
        return allGenerators
    }

    fun nestedGeneratorsRecursiveWithoutThis() = nestedGeneratorsRecursive().filter { it != this }

    private fun genRandomNestedGenerator(): Generator? {
        val queue = ArrayDeque<Generator>()
        val res = mutableListOf<Generator>()
        queue.add(this)
        while (queue.isNotEmpty()) {
            val el = queue.removeFirst()
            res.add(el)
        }
        return res.randomOrNull()
    }

    protected fun getAllGeneratorsBetween(start: Generator, end: Generator): List<Generator>? {
        val res = mutableListOf(mutableListOf(start))
        val queue = ArrayDeque<Generator>()
        queue.add(start)
        while (queue.isNotEmpty()) {
            val curGenerator = queue.removeFirst()
            if (curGenerator == end) break
            val nestedGenerators = curGenerator.nestedGenerators
            if (nestedGenerators.isEmpty()) continue
            val oldLists = res.removeIfAndReturnRemovedElements { it.last() == curGenerator }
            for (implementer in nestedGenerators) {
                queue.add(implementer)
                oldLists.forEach { res.add((it + listOf(implementer)).toMutableList()) }
            }
        }
        return res.find { it.last() == end }?.drop(1)?.dropLast(1)
    }

    private fun getAllGeneratorsBetween(currentPath: MutableList<Generator>, end: Generator) {

    }

    /**
     * @param type class token for type of property parameter this generator is
     * applicable to
     */
    protected constructor(type: Class<*>) : this(listOf(type))

    /**
     * Used for generators of primitives and their wrappers. For example, a
     * `Generator<Integer>` can be used for property parameters of type
     * `Integer` or `int`.
     *
     * @param types class tokens for type of property parameter this generator
     * is applicable to
     */
    init {
        this.types.addAll(types)
    }

    /**
     * @return class tokens for the types of property parameters this generator
     * is applicable to
     */
    fun types(): List<Class<*>> {
        return Collections.unmodifiableList(types)
    }

    /**
     * Tells whether this generator is allowed to be used for property
     * parameters of the given type.
     *
     * @param type type against which to test this generator
     * @return `true` if the generator is allowed to participate in
     * generating values for property parameters of `type`
     */
    open fun canRegisterAsType(type: Class<*>): Boolean {
        return true
    }

    /**
     *
     * This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.
     *
     * @return whether this generator has component generators, such as for
     * those generators that produce lists or maps
     * @see .addComponentGenerators
     */
    open fun hasComponents(): Boolean {
        return false
    }

    /**
     *
     * This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.
     *
     * @return how many component generators this generator needs
     * @see .addComponentGenerators
     */
    open fun numberOfNeededComponents(): Int {
        return 0
    }

    /**
     *
     * Adds component generators to this generator.
     *
     *
     * Some generators need component generators to create proper values.
     * For example, list generators require a single component generator in
     * order to generate elements that have the type of the list parameter's
     * type argument.
     *
     *
     * This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.
     *
     * @param newComponents component generators to add
     */
    open fun addComponentGenerators(newComponents: List<Generator>) {
        // do nothing by default
    }

    /**
     * @param typeParameters a list of generic type parameters
     * @return whether this generator can be considered for generating values
     * for property parameters that have the given type parameters in their
     * signatures
     */
    open fun canGenerateForParametersOfTypes(typeParameters: List<TypeParameter<*>>): Boolean {
        return true
    }

    /**
     *
     * Configures this generator using annotations from a given annotated
     * type.
     *
     *
     * This method considers only annotations that are themselves marked
     * with [GeneratorConfiguration].
     *
     *
     * By default, the generator will configure itself using this
     * procedure:
     *
     *  * For each of the given annotations:
     *
     *  * Find a `public` method on the generator named
     * `configure`, that accepts a single parameter of the
     * annotation type
     *  * Invoke the `configure` method reflectively, passing the
     * annotation as the argument
     *
     *
     *
     *
     * @param annotatedType a type usage
     * @throws GeneratorConfigurationException if the generator does not
     * "understand" one of the generation configuration annotations on
     * the annotated type
     */
    open fun configure(annotatedType: AnnotatedType?) {
        configureStrict(collectConfigurationJavaAnnotations(annotatedType))
    }

    /**
     * @param element an annotated program element
     */
    open fun configure(element: AnnotatedElement?) {
        configureLenient(collectConfigurationJavaAnnotations(element))
    }

    /**
     *
     * Supplies the available generators to this one.
     *
     *
     * This is intended for use only by junit-quickcheck itself, and not by
     * creators of custom generators.
     *
     * @param provided repository of available generators
     */
    open fun provide(provided: Generators) {
        repo = provided
    }

    fun isGeneratorContextInitialized() = this::generatorContext.isInitialized
    /**
     * Used by the framework to make a copy of the receiver.
     *
     * @return a copy of the receiver
     */
    open fun copy(): Generator {
        return Reflection.instantiate(javaClass).also {
            it.generatedUtModel = generatedUtModel
            it.generationState = generationState
            it.nestedGenerators = nestedGenerators.map { it.copy() }.toMutableList()
            if (isGeneratorContextInitialized()) {
                it.generatorContext = generatorContext
            }
            GeneratorConfigurator.configureGenerator(it, 95)
        }
    }

    /**
     * @return an access point for the available generators
     */
    protected fun gen(): Generators? {
        return repo
    }

    /**
     * @param random a source of randomness used when locating other
     * generators
     * @return an access point for the available generators
     */
    protected fun gen(random: SourceOfRandomness?): Generators {
        return repo!!.withRandom(random!!)
    }

    private fun collectConfigurationJavaAnnotations(
        element: AnnotatedElement?
    ): Map<Class<out JavaAnnotation>, JavaAnnotation> {
        if (element == null) return emptyMap()
        val configs = configurationAnnotationsOn(element)
        return configs.associateBy { it.annotationType() as Class<out JavaAnnotation> }
    }

    private fun configureStrict(byType: Map<Class<out JavaAnnotation>, JavaAnnotation>) {
        for ((key, value) in byType) configureStrict(key, value)
    }

    private fun configureStrict(
        annotationType: Class<out JavaAnnotation>,
        configuration: JavaAnnotation
    ) {
        configure(
            annotationType,
            configuration
        ) { ex: ReflectionException? ->
            throw GeneratorConfigurationException(
                String.format(
                    "Generator %s does not understand configuration annotation %s",
                    javaClass.name,
                    annotationType.name
                ),
                ex
            )
        }
    }

    private fun configureLenient(byType: Map<Class<out JavaAnnotation>, JavaAnnotation>) {
        for ((key, value) in byType) configureLenient(key, value)
    }

    private fun configureLenient(
        annotationType: Class<out JavaAnnotation>,
        configuration: JavaAnnotation
    ) {
        configure(annotationType, configuration) { ex: ReflectionException? -> }
    }

    private fun configure(
        annotationType: Class<out JavaAnnotation>,
        configuration: JavaAnnotation,
        exceptionHandler: (ReflectionException) -> Unit
    ) {
        var configurer: Method? = null
        try {
            configurer = Reflection.findMethod(javaClass, "configure", annotationType)
        } catch (ex: ReflectionException) {
            exceptionHandler(ex)
        }
        if (configurer != null) Reflection.invoke(configurer, this, configuration)
    }

    protected fun <T> withModification(block: () -> T): T {
        generationState = GenerationState.MODIFY
        return block.invoke().also { generationState = GenerationState.CACHE }
    }

    companion object {
        /**
         * @param parameter a generic type parameter
         * @param clazz a type
         * @return whether the type is compatible with the generic type parameter
         * @see .canGenerateForParametersOfTypes
         */
        protected fun compatibleWithTypeParameter(
            parameter: TypeParameter<*>,
            clazz: Class<*>?
        ): Boolean {
            return (parameter is WildcardTypeParameter
                    || parameter.type.isAssignableFrom(Types.forJavaLangReflectType(clazz)))
        }

        /**
         * Gives a list of the [GeneratorConfiguration] annotations present
         * on the given program element.
         *
         * @param element an annotated program element
         * @return what configuration annotations are present on that element
         */
        @JvmStatic
        protected fun configurationAnnotationsOn(element: AnnotatedElement): List<JavaAnnotation> =
            Reflection.allAnnotations(element)
                .filter {
                    it.annotationType().isAnnotationPresent(
                        GeneratorConfiguration::class.java
                    )
                }
    }
}